U/lynnej/timestep

rhiannonlynne opened the pull request:
Updated orbits.py to be much more sensible and compact on the logic of reading files. Also added CAR/CART options for standard DES files without headers. 

Updated makeLSSTobs.py to be more flexible when looking for columns in the opsim database. In particular, solarElong and rotskyPos may not be needed or present in the db, so make those more optional. 

Removed makeB612obs.py as this was just a copy of makeLSSTobs.py with arguments flipped to values that would make good defaults for B612 studies (at the time). The code is otherwise the same, so removed it as a duplicate that I didn't want to maintain.

rhiannonlynne commented on python/lsst/sims/movingObjects/orbits.py:
```diff
@@ -203,76 +206,88 @@ def assignSed(self, orbits, randomSeed=None):
-        """Read orbits from a file, generating a pandas dataframe containing columns matching
-        dataCols, for the appropriate orbital parameter format (currently accepts COM or KEP formats).
+        """Read orbits from a file, generating a pandas dataframe containing columns matching dataCols,
+        for the appropriate orbital parameter format (currently accepts COM, KEP or CAR/CART formats).
```
The 'CAR' keyword isn't something that I found documented anywhere (presumably it's in the DES data format standard, but the doc on that is very hard to find .. I know I saw it at some point, but have been unable to find it ever since), so I just started using "CART" for "Cartesian" instead of keeping to the 3-characters used for COMetary and KEPlerian. This has now propagated to sbpy, at least in dictionary keys ... 
https://github.com/NASA-Planetary-Science/sbpy/search?q=CART&unscoped_q=CART

So I guess I'd leave it and support both? It does mean that 'CART' may propagate further, not sure how bad that would be.

mjuric replied:
Looks like the standard doc (http://adams.dm.unipi.it/orbfit/PS_DES_v2.02.pdf -- is that it) specifies "CAR" (see page 7); may be best to stop the propagation while it didn't go too far (and ping @mommermi about deprecating it in `sbpy`). Reducing complexity is good :).

mommermi replied:
Interesting, I didn't even know that 'CAR' existed and I took 'CART' from Lynne's example notebook. I'm perfectly fine with switching to 'CAR'...

mjuric replied:
Another data point is that command-line OpenOrb doesn't understand `CART`.

rhiannonlynne replied:
Ok, I'm convinced and I'll start removing CART from here.

rhiannonlynne commented on python/lsst/sims/movingObjects/orbits.py:
```diff
@@ -203,76 +206,88 @@ def assignSed(self, orbits, randomSeed=None):
+            skiprows = -1
             # Figure out whether the header is in the first line, or if there are rows to skip.
             # We need to do a bit of juggling to do this before pandas reads the whole orbit file though.
             file = open(orbitfile, 'r')
```
Yes, good point. (old code, pre-context-manager, not upgraded).

rhiannonlynne commented on python/lsst/sims/movingObjects/orbits.py:
```diff
@@ -203,76 +206,88 @@ def assignSed(self, orbits, randomSeed=None):
+                          'H', 'epoch', 'INDEX', 'N_PAR', 'MOID', 'COMPCODE')
+            # First use names_COM, and then change if required.
+            orbits = pd.read_csv(orbitfile, delim_whitespace=True, header=None, names=names_COM)
+            if orbits['FORMAT'][0] == 'KEP':
```
There isn't really a reason to not support mixed format except for not keeping this column for every object and then not checking the value when doing the conversion to/from oorb element format. For simplicity's sake, it's not -- but this is mostly because it's just not a use-case that's come up (usually people have elements that are all one or the other, for one type of object). 

My inclination would be to do the check and then throw an error .. if you think it'd important to keep the capability to handle mixed entries, then there are some more changes to make.

mjuric replied:
Agreed that it's not important! But let's throw an exception (otherwise the code will silently do the wrong thing).

rhiannonlynne commented on bin.src/makeLSSTobs.py:
```diff
@@ -13,21 +13,49 @@
 
 
-def readOpsim(opsimfile, constraint=None, dbcols=None):
+def readOpsim(opsimfile, constraint=None, footprint='camera', dbcols=None):
```
Hmm. I suppose. It's really only there so that I could decide if 'rotSkyPos' was a required column or not (I'd like to just say it's required, but Bryce's old subset of a opsim database didn't have it, so I couldn't just do that).

rhiannonlynne replied:
Added documentation to makeLSSTobs.py methods.

rhiannonlynne commented on bin.src/makeLSSTobs.py:
```diff
@@ -13,21 +13,49 @@
+        except ValueError:
+            failed_cols.append(col)
+    if len(failed_cols) > 0:
+        print('Could not find columns %s in the database, so leaving those out for now.' % failed_cols)
```
If the message comes up, the user doesn't really have to *deal* with it, but they might like to know about it (generally because it's likely to be either unexpected to them, or it's expected and then they can ignore it). I guess I'm assuming the user kind of knows what's in their database and if something unexpected shows up here (couldn't find some column that they thought was there), they might want to check into it a bit more.

I haven't used proper logging .. you're probably right that this is what should be happening, if only for better redirect capability. 

Actual problematic errors anywhere in the code will raise exceptions and halt execution. Anything else is just INFO level.  To me, DEBUG seems like you wouldn't have it on in usual execution?

mjuric replied:
I'm torn between making this an INFO (with a default logging level of WARN) or WARN (but raise the default logging level to ERROR). Either way, I think the users shouldn't have it displayed by default -- if they encounter an error they could increase the log level and get this extra information.

Why: My guess is the vast majority of users won't know what this message is about -- they'll be using the code as a black box, with databases we've given them (with maybe an ε more who have read the documentation :-)). A good rule of thumb may be to write messages to the console if and only if the user should act on them. Otherwise they get confused and either don't trust the result ("oh my, look at all these messages, this can't be good!"), or stop looking at the messages all together and fail to notice the really important ones that indicate something /is/ fishy with the result.

Since Python's default logging level is WARN, that would make most of these types of messages INFO. Or one could configure the logger to raise the console output level to ERROR or CRITICAL by default, so these types of things could be WARN.

Here are [Python's official guidelines](https://docs.python.org/3/howto/logging.html#when-to-use-logging). LSST pipelines [by default log INFO to the console](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/log.html), though they're not intended for broad use (and generally dump way too much output to the console). 

PS: Or maybe even DEBUG -- if we follow the Python guidelines, these messages will only be used when diagnosing problems, and that's what DEBUG is for (INFO is for "confirmation that things are working as expected" whereas DEBUG is for "Detailed information, typically of interest only when diagnosing problems.")

rhiannonlynne replied:
I guess I view these as "Info" - "confirmation that things are working as expected".
I did set it so that info comments were sent to the console .. I guess I feel there's not that many of them and they are helpful for users to know what's going on when.

rhiannonlynne replied:
I also updated any 'print' statements to logging messages.

rhiannonlynne commented on bin.src/makeLSSTobs.py:
```diff
@@ -13,21 +13,49 @@
+        more_cols.remove(col)
+    cols = min_cols + more_cols
+    cols = list(set(cols))
+    print('Querying for columns:\n %s' % (cols))
```
So 'debug' level stuff is informational for the user? (I'm assuming the user would like to see what is being pulled out of the opsim database, as this is what goes into their obs file and is then available to MAF).

mjuric replied:
But won't they know what's been pulled out by looking at the output?

Debug is for diagnosing problems -- my guess was that you had this print statement so the user could see what was queried if they look at the output and don't find an expected column (for example).

rhiannonlynne replied:
They would know that by looking at their output, yes .. but that may come out quite a bit after the start.

mjuric commented on bin.src/makeLSSTobs.py:
```diff
@@ -13,21 +13,49 @@
+    # Go ahead and query for all of the observations.
+    simdata = opsdb.fetchMetricData(cols, sqlconstraint=constraint)
     opsdb.close()
     print("Queried data from opsim %s, fetched %d visits." % (opsimfile, len(simdata)))
```
I take this comment back -- this is effectivelly a progress message in a console program, so `print` is [fine](https://docs.python.org/3/howto/logging.html#when-to-use-logging).

rhiannonlynne commented on python/lsst/sims/movingObjects/orbits.py:
```diff
@@ -203,76 +206,88 @@ def assignSed(self, orbits, randomSeed=None):
-                    valuesheader = values
-            skiprows -= 1
+                    headervalues = values
             file.close()
```
Okay ... so I think this is a good idea, but I would kind of just like to resolve the PR now and file this as an issue/upgrade for later.

rhiannonlynne commented:
I think I addressed most of the issues, and the unit test was expanded to look at more types of inputs. I did not modify the readOrbits methods to read from stringIO type objects. Something to look at in the future - a good idea, but I'd like to get this merged and tackle that separately.
